src/util/buffer.rs

```diff
@@ -0,0 +1,284 @@
+            *self = Buffer::allocate(cap)
+        } else {
+            // We need to set the capacity to 0 because if the capacity
+            // overflows unwinding is triggered, which if we don't
+            // change the capacity would try to free empty().
+            let old_cap = mem::replace(&mut self.cap, 0);
+            let buffer = mem::replace(&mut self.buffer, empty());
+
+            self.buffer = unsafe {
+                reallocate(buffer, NonZero::new(old_cap),
+                           NonZero::new(cap))
+            };
+            self.cap = cap;
+        }
+    }
+
+    /// Get the current capacity of the Buffer.
+    ///
+    /// ```
+    /// # use collect::util::buffer::Buffer;
+    ///
+    /// let buffer: Buffer<usize> = Buffer::allocate(128);
+    /// assert_eq!(buffer.capacity(), 128);
+    /// ```
+    pub fn capacity(&self) -> usize {
+        self.cap
+    }
+}
+
+#[unsafe_destructor]
+impl<T> Drop for Buffer<T> {
+    /// Safety Note:
+    ///
+    /// The Buffer will *only* deallocate the contained memory. It will
+    /// *not* run any destructors on data in that memory.
+    fn drop(&mut self) {
+        if self.cap == 0 { return }
+        unsafe { deallocate(self.buffer, NonZero::new(self.cap)) }
+    }
+}
+
+impl<T> Deref for Buffer<T> {
+    type Target = *mut T;
+
+    fn deref(&self) -> &*mut T {
+        &*self.buffer
+    }
+}
+
+// Typed Allocation Utilities
+//
+// Unlike std::rt::heap these check for zero-sized types, capacity overflow,
+// oom etc. and calculate the appropriate size and alignment themselves.
+
+unsafe fn allocate<T>(cap: NonZero<usize>) -> NonZero<*mut T> {
+    if mem::size_of::<T>() == 0 { return empty() }
+
+    // Allocate
+    let ptr = heap::allocate(allocation_size::<T>(cap), mem::align_of::<T>());
+
+    // Check for allocation failure
+    if ptr.is_null() { ::alloc::oom() }
+
+    NonZero::new(ptr as *mut T)
+}
+
+unsafe fn reallocate<T>(ptr: NonZero<*mut T>,
+                        old_cap: NonZero<usize>,
+                        new_cap: NonZero<usize>) -> NonZero<*mut T> {
+    if mem::size_of::<T>() == 0 { return empty() }
+
+    let old_size = unchecked_allocation_size::<T>(old_cap);
+    let new_size = allocation_size::<T>(new_cap);
+
+    // Reallocate
+    let new = heap::reallocate(*ptr as *mut u8, old_size, new_size, mem::align_of::<T>());
+
+    // Check for allocation failure
+    if new.is_null() {
+        // Cleanup old buffer.
+        deallocate(ptr, old_cap);
+
+        ::alloc::oom()
+    }
+
+    NonZero::new(new as *mut T)
+}
+
+unsafe fn deallocate<T>(ptr: NonZero<*mut T>, cap: NonZero<usize>) {
+    if mem::size_of::<T>() == 0 { return }
+
+    let old_size = unchecked_allocation_size::<T>(cap);
+
+    heap::deallocate(*ptr as *mut u8, old_size, mem::align_of::<T>())
+}
+
+fn allocation_size<T>(cap: NonZero<usize>) -> usize {
+    mem::size_of::<T>().checked_mul(*cap).expect("Capacity overflow")
```

Comment:
So this is kinda a Problem Rust Has. Technically overflowing a uint isn't sufficient. Anything bigger than `int::MAX` is going to cause some UB because pointer offsets will overflow. Technically possible to create such a thing without OOM on e.g. x32 (32-bit mode binary compiled for a 64-bit platform).

This isn't something you really need to address. It just continues to distress me. :crying_cat_face: